### PR TITLE
disabling-doubleposts feature

### PR DESCRIPTION
**Pull request by Risers**
Doubleposting is disabled server-side and user-side. Users can doublepost 3 days after last post.
User that created topic can doublepost every 24h.

### CODE REVIEW

File: app/Models/Forum/Topic.php [hunk 1]
```diff
@@ -425,4 +426,30 @@ public function deleteWithCover()
...
         $this->delete();
     }
+
+    public function isADoublePost(User $user)
+    {
+        if ($user === null) {
+            return false;
+        }
+        if ($user->user_id === $this->topic_poster) {
+            if (Carbon::now()->subdays(1) > $this->topic_last_post_time) {
```
Comment (peppy) [inline app/Models/Forum/Topic.php]: this should be config level

File: app/Models/Forum/Topic.php [hunk 2]
```diff
@@ -425,4 +426,30 @@ public function deleteWithCover()
...
+                return false;
+            }
+            if ($user->user_id !== $this->topic_last_poster_id) {
+                return false;
+            }
+
+            return true;
+        } else {
+            if (Carbon::now()->subdays(3) > $this->topic_last_post_time) {
```
Comment (peppy) [inline app/Models/Forum/Topic.php]: this should be config level

**Comment by peppy** [comment]
what's with all the file permission changes? might want to check your git setup.

**Comment by LiquidPL** [comment]
This gets created on a clean vagrant provision iirc

**Comment by Sakisan** [comment]
I also get these modifications automatically right after the installation by vagrant.
```
~/projects/osu-web/code on master !
➜ gss
 M bootstrap/cache/.keep
 M storage/.gitignore
 M storage/app/.gitignore
 M storage/clockwork/.gitignore
 M storage/framework/.gitignore
 M storage/framework/cache/.gitignore
 M storage/framework/sessions/.gitignore
 M storage/framework/views/.gitignore
 M storage/htmlpurifier/.keep
 M storage/logs/.gitignore
```

No one noticed this before?

**Comment by LiquidPL** [comment]
I did, but I didn't really care to bring that up

**Comment by Sakisan** [comment]
I guess the quick fix is to just let these files be merged. I don't think that would be a problem.

**Comment by Risers** [comment]
Kinda up?

File: app/Models/Forum/Topic.php [hunk 3]
```diff
@@ -425,4 +426,30 @@ public function deleteWithCover()
 
         $this->delete();
     }
+
+    public function isADoublePost(User $user)
```
Comment (nanaya) [inline app/Models/Forum/Topic.php]: should be `isDoublePostBy`

File: app/Models/Forum/Topic.php [hunk 4]
```diff
@@ -425,4 +426,30 @@ public function deleteWithCover()
...
+    {
+        if ($user === null) {
+            return false;
+        }
+        if ($user->user_id === $this->topic_poster) {
+            if (Carbon::now()->subhours(config('osu.forum.hours_before_topic_author_doublepost')) > $this->topic_last_post_time) {
+                return false;
+            }
+            if ($user->user_id !== $this->topic_last_poster_id) {
```
Comment (nanaya) [inline app/Models/Forum/Topic.php]: this should be checked before the outer conditional

File: app/Models/Forum/Topic.php [hunk 5]
```diff
@@ -425,4 +426,30 @@ public function deleteWithCover()
 
         $this->delete();
     }
+
+    public function isADoublePost(User $user)
+    {
+        if ($user === null) {
+            return false;
+        }
+        if ($user->user_id === $this->topic_poster) {
```
Comment (nanaya) [inline app/Models/Forum/Topic.php]: this conditional should just assign `$minimumHours` and do the actual check afterward.

File: resources/assets/coffee/forum/post-box.coffee [hunk 6]
```diff
@@ -1,17 +1,13 @@
 ###
 Copyright 2015 ppy Pty. Ltd.
-
```
Comment (nanaya) [inline resources/assets/coffee/forum/post-box.coffee]: ??

File: resources/assets/less/bem/forum-post.less [hunk 7]
```diff
@@ -383,4 +385,24 @@
...
+        position:absolute;
+        top: 0;
+        left: 0;
+        height: 100%;
+        width: 100%;
+        text-align: center;
+        font-size: 30px;
+        color: white;
+        padding-top: 100px;
+	      filter:alpha(opacity=50);
```
Comment (nanaya) [inline resources/assets/less/bem/forum-post.less]: broken indentation, also there's no need for this ie crap

**Comment by nanaya** [comment]
- what's with so many "style fixes" commit?
- don't change permission of unrelated files. And, no, those files should not be executable. 
- once activated (and is double post) the box can't be hidden since close button is obstructed by the overlay
- if the user has slow connection, he will be able to write something and then later the overlay appears
- every reply box activation triggers network round trip. Especially noticeable when scrolling back and forth from the bottom of the page while writing a reply

**Comment by Risers** [comment]
Thanks

**Comment by Risers** [comment]
Done

File: app/Models/Forum/Topic.php [hunk 8]
```diff
@@ -403,4 +404,27 @@ public function deleteWithCover()
...
+    public function isDoublePostBy(User $user)
+    {
+        if ($user === null) {
+            return false;
+        }
+        if ($user->user_id !== $this->topic_last_poster_id) {
+            return false;
+        }
+        if ($user->user_id === $this->topic_poster) {
+            if (Carbon::now()->subhours(config('osu.forum.authorDoublePostTime')) > $this->topic_last_post_time) {
```
Comment (nanaya) [inline app/Models/Forum/Topic.php]:
```php
if ($user->user_id === $this->topic_poster) {
    $minTime = config(...);
} else {
    $minTime = config(...);
}

return Carbon::now()->subhours(...) < $this->...;
```

File: config/osu.php [hunk 9]
```diff
@@ -26,7 +26,8 @@
     'forum' => [
         'admin_forum_id' => intval(env('ADMIN_FORUM_ID', 28)),
         'help_forum_ids' => array_map('intval', explode(' ', env('HELP_FORUM_IDS', '4 5 29 30 101'))),
-
+        'doublePostTime' => 72,
```
Comment (nanaya) [inline config/osu.php]: `=> ['normal' => 72, 'author' => 24]`?

**Comment by nanaya** [comment]
Just make it like this



In `resources/views/forum/topics/show.blade.php` around line 137:

```php
<div class="forum-post__content forum-post__content forum-post__content--edit-bar">
    @if (priv_check('ForumTopicReply', $topic)->can())
        @include("forum.topics._post_box_footer", ["submitText" => trans("forum.topic.post_reply")])
    @else
        <span>
            <i class="fa fa-warning"></i>
            {{ priv_check('ForumTopicReply', $topic)->message() }}
        </span>
    @endif
</div>
```

And add relevant check in `app/Libraries/OsuAuthorize.php`, function `checkForumTopicReply`.

**Comment by Risers** [comment]
Done

File: config/osu.php [hunk 10]
```diff
@@ -29,6 +29,10 @@
     'forum' => [
         'admin_forum_id' => intval(env('ADMIN_FORUM_ID', 28)),
         'help_forum_ids' => array_map('intval', explode(' ', env('HELP_FORUM_IDS', '4 5 29 30 101'))),
+        'doublePostTime' => [
```
Comment (nanaya) [inline config/osu.php]: everything else here is `snake_cased` so this one should be as well.

**Comment by Risers** [comment]
Done

**Pull request merged**